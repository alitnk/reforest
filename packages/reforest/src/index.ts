export { createTreeProvider, stringifyTreeMap } from "./server"
export { useIndex, useIndexedChildren } from "./use-indexed-children"
export type { TreeState } from "./use-tree"
export { useTree, useTreeData, useTreeState } from "./use-tree"
export { cleanAndSortTree, mapToChildren, parseIndexPath, sortMapByIndexPath } from "./utils"
