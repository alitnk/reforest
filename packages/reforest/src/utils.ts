import * as React from "react"
import { arrayToTree } from "performant-array-to-tree"

export const DATA_ID = "__REFOREST_DATA__"

export const isServer = typeof window === "undefined"

export const useIsomorphicLayoutEffect = isServer ? React.useEffect : React.useLayoutEffect

/**
 * Parses a numerical dot-separated string as an index path.
 *
 * @example
 * parseIndexPath('0.10.2') -> [0, 10, 2]
 */
export function parseIndexPath(indexPathString: string) {
  return indexPathString.split(".").map((index) => parseInt(index, 10))
}

/** Recursive function that removes "id" and "parentId" keys and returns each indexed data. */
export function cleanAndSortTree(tree: any) {
  if (tree.children?.length > 0) {
    /** Sort children by the index path. */
    tree.children.sort((a, b) => parseFloat(a.indexPathString) - parseFloat(b.indexPathString))

    return {
      ...tree.data,
      children: tree.children.map(cleanAndSortTree),
    }
  }

  return tree.data
}

/** Builds a tree from a Map of data collected in useTree. */
export function mapToTree(dataMap: Map<string, any>) {
  const parsedValues = Array.from(dataMap.values()).map((data) => {
    const parentIndexPathString = parseIndexPath(data.indexPathString).slice(0, -1).join(".")

    return {
      data,
      parentId: parentIndexPathString,
      id: data.indexPathString,
    }
  })
  const tree = arrayToTree(parsedValues, { dataField: null })
  const cleanedTree = cleanAndSortTree({ children: tree })

  return cleanedTree
}

/** Sorts a map by an indexPathString property. */
export function sortMapByIndexPath(treeMap: Map<string, any>) {
  const sortedEntries = Array.from(treeMap.entries()).sort(
    (a, b) => parseFloat(a[1].indexPathString) - parseFloat(b[1].indexPathString)
  )

  return new Map(sortedEntries)
}
