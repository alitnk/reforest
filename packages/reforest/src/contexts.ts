import * as React from "react"

import { DATA_ID, isServer } from "./utils"

export const MaxIndexContext = React.createContext<number[]>([])

MaxIndexContext.displayName = "MaxIndexContext"

export const IndexContext = React.createContext<string | null>(null)

IndexContext.displayName = "IndexContext"

/** Gets the initial computed data from the injected script tag. */
export function getInitialComputedData() {
  let serverEntries: [string, any][] = []

  /** Hydrate data if available in document head. */
  if (!isServer) {
    const serverJSON = document.getElementById(DATA_ID)?.innerHTML

    if (serverJSON) {
      const serverComputedData = JSON.parse(serverJSON)

      if (serverComputedData) {
        serverEntries = Object.entries(serverComputedData)
      }
    }
  }

  return new Map(serverEntries)
}

export const ComputedDataContext = React.createContext<Map<string, any> | null>(
  getInitialComputedData()
)

ComputedDataContext.displayName = "ComputedDataContext"

export type TreeStateContextValue<ComputedData extends any = any> = {
  treeMap: Map<string, any>
  subscribeTreeData: (key: string, value: any) => () => void
} & {
  computed: ComputedData | null
}

export const TreeStateContext = React.createContext<TreeStateContextValue | null>(null)
