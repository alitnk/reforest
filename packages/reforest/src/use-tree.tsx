import * as React from "react"
import create from "zustand"

import { useIndex } from "./use-indexed-children"
import { isServer, mapToTree, useIsomorphicLayoutEffect } from "./utils"

let globalTimeoutId: ReturnType<typeof setTimeout>

/** Global store that manages all tree state. */
export const useTreeStore = create<{
  needsUpdate: boolean
  treeMap: Map<string, any>
  subscribeTreeData: (key: string, value: any) => () => void
  clearTreeData: () => void
}>((set, get) => ({
  needsUpdate: false,
  treeMap: new Map(),
  subscribeTreeData: (id, data) => {
    const { treeMap, needsUpdate } = get()

    treeMap.set(id, data)

    /** Keep clearing timeout until the last hook renders. */
    clearTimeout(globalTimeoutId)

    if (needsUpdate === false) {
      set({ needsUpdate: true })
    }

    globalTimeoutId = setTimeout(() => {
      set({ needsUpdate: false })
    })

    return () => {
      treeMap.delete(id)
    }
  },
  clearTreeData: () => {
    set({ treeMap: new Map() })
  },
}))

/**
 * Subscribes data to the global tree context and returns a unique id that can
 * be used for fetching computed data in useComputedData.
 */
export function useTreeData(data: any) {
  const treeId = React.useId()
  const index = useIndex()
  const indexPathString = index ? index.indexPathString : null

  /** Capture the initial data when rendering on the server. */
  if (isServer) {
    useTreeStore
      .getState()
      .subscribeTreeData(treeId, Object.assign({ treeId, indexPathString }, data))
  }

  useIsomorphicLayoutEffect(() => {
    return useTreeStore
      .getState()
      .subscribeTreeData(treeId, Object.assign({ treeId, indexPathString }, data))
  }, [treeId, indexPathString, data])

  return treeId
}

/** Subscribe to tree updates. */
export function useTreeEffect(
  /** Updates when any tree data is changed. */
  onUpdate: (tree: ReturnType<typeof mapToTree>) => void | (() => void),

  /** Dependencies for the effect. */
  dependencies: any[] = []
) {
  const onUpdateRef = React.useRef<typeof onUpdate>(onUpdate)
  const cleanupRef = React.useRef<ReturnType<typeof onUpdate> | null>(null)
  const previousStringifiedTree = React.useRef("")

  useIsomorphicLayoutEffect(() => {
    onUpdateRef.current = onUpdate
  })

  useIsomorphicLayoutEffect(() => {
    function handleTreeUpdate(state: any) {
      let treeData = {}

      state.treeMap.forEach((data, key) => {
        treeData[key] = data
      })

      const nextStringifiedTree = JSON.stringify(treeData)

      if (previousStringifiedTree.current !== nextStringifiedTree) {
        const tree = mapToTree(state.treeMap)

        if (tree) {
          cleanupRef.current = onUpdateRef.current(tree)
        }

        previousStringifiedTree.current = nextStringifiedTree
      }
    }

    /** Build initial tree once children have rendered. */
    handleTreeUpdate(useTreeStore.getState())

    /** Subscribe to future updates to tree. */
    const cleanupSubscription = useTreeStore.subscribe(handleTreeUpdate)

    return () => {
      cleanupSubscription()
      cleanupRef.current?.()
      previousStringifiedTree.current = ""
    }
  }, dependencies)
}
