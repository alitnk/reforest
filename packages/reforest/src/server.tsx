import * as React from "react"
import { suspend } from "suspend-react"

import { ComputedDataContext, TreeMapContext } from "./contexts"
import { isServer, sortMapByIndexPath } from "./utils"

const DATA_ID = "__REFOREST_DATA__"

/** Stringify a tree map. */
export function stringifyTreeMap(treeMap: Map<string, any>) {
  let allData = {}

  treeMap.forEach((data, key) => {
    allData[key] = data
  })

  return JSON.stringify(allData)
}

/** Wraps a tree with a Map for gathering indexed data on the server. */
export function createTreeProvider(initialEntries?: [string, any][]) {
  const treeComputedData = new Map<string, any>(initialEntries)

  function TreeProvider(props: { children: React.ReactNode }) {
    return (
      <ComputedDataContext.Provider value={treeComputedData}>
        {props.children}
      </ComputedDataContext.Provider>
    )
  }

  return {
    TreeProvider,
    treeComputedData,
    stringifyTreeComputedData: () => stringifyTreeMap(treeComputedData),
    getInitializerScript: () =>
      `<script id="${DATA_ID}" type="application/json">${stringifyTreeMap(
        treeComputedData
      )}</script>`,
  }
}

/** Track initial component renders. */
let globalResolves: any[] = []
let globalTimeoutId: ReturnType<typeof setTimeout>

/** Compute data based on all collected tree data. */
export function useServerComputedData<TreeValue extends any, ComputedTreeValue extends any>(
  treeId: string,
  computeData?: (treeMap: Map<string, TreeValue>, treeId: string) => ComputedTreeValue
) {
  const treeMap = React.useContext(TreeMapContext)
  const treeComputedData = React.useContext(ComputedDataContext)

  /** Use Suspense to re-render the component before committing the final props on the server. */
  const isServerWithComputedData = isServer && treeMap && computeData !== undefined
  let serverComputedData: any = null

  if (isServerWithComputedData) {
    serverComputedData = suspend(() => {
      return new Promise((resolve) => {
        /** Keep clearing timeout until the last component renders. */
        clearTimeout(globalTimeoutId)

        /** Store all of the promises to compute. */
        globalResolves.push(() => {
          const sortedTreeMap = sortMapByIndexPath(treeMap)
          const computedData = computeData(sortedTreeMap, treeId)

          resolve(computedData)
        })

        /** Push to the end of the event stack to allow all leaf components to initially render. */
        globalTimeoutId = setTimeout(() => {
          /** Resolve all of the leaf promises now that we have stored and computed all data. */
          globalResolves.forEach((resolve) => resolve())
          globalResolves = []
        })
      })
    }, [treeId])

    /** Store computed data so it can be injected on the server. */
    treeComputedData?.set(treeId, serverComputedData)
  } else {
    serverComputedData = treeComputedData?.get(treeId)
  }

  return serverComputedData
}
