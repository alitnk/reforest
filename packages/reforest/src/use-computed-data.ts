import * as React from "react"
import { suspend } from "suspend-react"

import { ComputedDataContext, TreeStateContext } from "./contexts"
import { isServer, sortMapByIndexPath, useIsomorphicLayoutEffect } from "./utils"

/** Track initial component renders. */
let globalResolves: any[] = []
let globalTimeoutId: ReturnType<typeof setTimeout>

/** Compute data based on all collected tree data. */
export function useComputedData<ComputedData extends any>(
  computeData: (
    treeMap: Map<string, { generatedId: string; indexPathString: string } & Record<string, any>>
  ) => ComputedData,
  dependencies: any[] = []
) {
  const treeState = React.useContext(TreeStateContext)
  const treeComputedData = React.useContext(ComputedDataContext)
  const generatedId = React.useId()

  if (treeState === null) {
    throw new Error("useComputedData must be used within a component that uses useTree.")
  }

  /** Use Suspense to re-render the component before committing the final props on the server. */
  const isServerWithComputedData = isServer && computeData !== undefined
  let serverComputedData: any = null

  if (isServerWithComputedData) {
    serverComputedData = suspend(() => {
      return new Promise((resolve) => {
        /** Keep clearing timeout until the last component renders. */
        clearTimeout(globalTimeoutId)

        /** Store all of the promises to compute. */
        globalResolves.push(() => {
          const sortedTreeMap = sortMapByIndexPath(treeState.treeMap)
          const computedData = computeData(sortedTreeMap)

          resolve(computedData)
        })

        /** Push to the end of the event stack to allow all leaf components to initially render. */
        globalTimeoutId = setTimeout(() => {
          /** Resolve all of the leaf promises now that we have stored and computed all data. */
          globalResolves.forEach((resolve) => resolve())
          globalResolves = []
        })
      })
    }, [generatedId])

    /** Store computed data so it can be injected on the server. */
    treeComputedData?.set(generatedId, serverComputedData)
  } else {
    serverComputedData = treeComputedData?.get(generatedId)
  }

  const clientComputedData = null

  // const snapshot = useSnapshot(treeState)
  // const clientComputedData = React.useMemo(
  //   () => (computeData ? computeData(sortMapByIndexPath(snapshot.treeMap)) : null),
  //   dependencies.concat(snapshot)
  // )

  // useIsomorphicLayoutEffect(() => {
  //   treeComputedData?.set(generatedId, clientComputedData)
  //   return () => {
  //     treeComputedData?.delete(generatedId)
  //   }
  // }, [clientComputedData, generatedId])

  return clientComputedData || serverComputedData
}
