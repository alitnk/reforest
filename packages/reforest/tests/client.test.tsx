import * as React from "react"
import { waitFor, render } from "@testing-library/react"
import "@testing-library/jest-dom"

import { useIndexedChildren, useIndex, useTreeData, useTreeEffect } from "../src"

import { App } from "./App"

test("renders a simple list of items with the correct indexes", async () => {
  function Item({ children, value }: { children: React.ReactNode; value: string }) {
    const indexedChildren = useIndexedChildren(children)
    const index = useIndex()

    useTreeData(React.useMemo(() => ({ value }), [value]))

    return <div data-testid={index?.indexPathString}>{indexedChildren}</div>
  }

  function ItemList({ children }: { children: React.ReactNode }) {
    const indexedChildren = useIndexedChildren(children)

    return indexedChildren
  }

  await waitFor(() => {
    const { queryByTestId } = render(
      <React.Suspense fallback={null}>
        <ItemList>
          <Item value="apple">Apple</Item>
          <Item value="orange">Orange</Item>
          <Item value="banana">Banana</Item>
        </ItemList>
      </React.Suspense>
    )

    expect(queryByTestId("1")).toHaveTextContent("Orange")
  })
})

test("renders a complex list of items with the correct indexes", async () => {
  const handleTreeUpdate = jest.fn()

  function Item({ children, value }: { children: React.ReactNode; value: string }) {
    const indexedChildren = useIndexedChildren(children)
    const index = useIndex()

    useTreeData(React.useMemo(() => ({ value }), [value]))

    return <div data-testid={index?.indexPathString}>{indexedChildren}</div>
  }

  function ItemList({ children }: { children: React.ReactNode }) {
    const indexedChildren = useIndexedChildren(children)

    useTreeEffect(handleTreeUpdate)

    return indexedChildren
  }

  await waitFor(() => {
    const { queryByTestId } = render(
      <React.Suspense fallback={null}>
        <ItemList>
          <Item value="apples">
            Apples
            <Item value="fuji">Fuji</Item>
            <Item value="gala">Gala</Item>
            <Item value="honeycrisp">Honeycrisp</Item>
          </Item>
          <Item value="apples">
            Oranges
            <Item value="mandarin">Mandarin</Item>
            <Item value="naval">Naval</Item>
            <Item value="tangerine">Tangerine</Item>
          </Item>
          <Item value="pears">
            Pears
            <Item value="anjou">Anjou</Item>
            <Item value="asian">Asian</Item>
            <Item value="bosc">Bosc</Item>
          </Item>
        </ItemList>
      </React.Suspense>
    )

    expect(queryByTestId("2.3")).toHaveTextContent("Bosc")

    expect(handleTreeUpdate).toHaveBeenCalledTimes(1)

    expect(handleTreeUpdate).toMatchSnapshot()
  })
})

test("renders and computes layout", () => {
  const { container, findByText } = render(<App />)

  findByText("Box")

  expect(container.firstChild).toMatchSnapshot()
})
