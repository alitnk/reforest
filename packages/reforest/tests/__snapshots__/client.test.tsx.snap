// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and computes layout 1`] = `
<div>
  <div
    style="width: 200px; height: 133.33333333333334px; top: 201px; left: 134.33333333333334px;"
  >
    Box
  </div>
</div>
`;
